### PR TITLE
Correctly expand environment variables at the right time

### PR DESCRIPTION
**Pull request by andreia-oca**
# Pull Request Template

## Type of change

-   [ ] 🍕 New feature
-   [x] 🐛 Bug Fix
-   [ ] 🔥 Breaking change
-   [ ] 🧑‍💻 Improvement
-   [ ] 📝 Documentation Update

## Checklist

-   [x] My code follows the contributor guidelines of this project;
-   [ ] I have updated the documentation;
-   [ ] I have added tests;
-   [x] New and existing unit tests pass locally with my changes;

### CODE REVIEW

**Review by cristim67 — COMMENTED** [review]
i don't think you search property the env variables.

i have this project tree:

```
.venv(base) cristimiloiu@MacBook-Pro-Cristi nestjs-react-getting-started % tree  -L 2 -a
.
├── backend
│   ├── .env
│   ├── ...
├── frontend
│   ├── ...
└── genezio.yaml

17 directories, 26 files
```

and logs:

```
.venv(base) cristimiloiu@MacBook-Pro-Cristi nestjs-react-getting-started % genezio deploy
09:04:26:501 DEBUG      index.js:24     Sentry not initialized Cannot find module '/Users/cristimiloiu/Desktop/workspace/work/genezio/node_modules/@sentry/profiling-node/lib/sentry_cpu_profiler-darwin-arm64-131.node'
Require stack:
- /Users/cristimiloiu/Desktop/workspace/work/genezio/node_modules/@sentry/profiling-node/lib/index.js
Version: 3.2.2
09:04:26:505 DEBUG      command.js:17   The following components will be build and deployed: nest, classic
09:04:26:505 DEBUG      command.js:39   Deploying Nest.js app
✔ We detected that SECRET are not set remotely. Do you want us to set them for you? Yes
09:04:29:907 DEBUG      utils.js:643    Uploading environment variables [{"name":"SECRET","value":"test"}] from /Users/cristimiloiu/Desktop/workspace/work/nestjs-react-getting-started/backend/.env.
09:04:29:908 DEBUG      utils.js:37     No services found in the configuration file.
09:04:29:908 DEBUG      utils.js:133    Attempting to install dependencies with npm 
Dependencies installed successfully with command: npm install

 Error  File /Users/cristimiloiu/Desktop/workspace/work/nestjs-react-getting-started/.env does not exist. Please provide the correct path using `--env <envFile>`.
```

looks like you find the envs, but you raise an error

**Review by cristim67 — COMMENTED** [review]
if you expect that .env to be in route, why don't you put it for the backend? i think it's a problem for ssr.

i have this project tree:

```
.venv(base) cristimiloiu@MacBook-Pro-Cristi nestjs-react-getting-started % tree  -L 2 -a
.
├── backend
│   ├── ...
├── frontend
│   ├── ...
├── .env
└── genezio.yaml

17 directories, 26 files
```

and the logs:

```
09:09:07:962 DEBUG-nest:genezioAdapter.js:95████Done uploading the content to S3 for function: function-nest..


✔ Checking your credentials
09:09:07:966 DEBUG      deployCode.js:54        Deploy request sent with body: {"options":{"architecture":"x86_64"},"classes":[],"functions":[{"name":"function-nest","language":"js","fileName":"genezioDeploy.zip","entryFile":"main.js","persistent":false}],"projectName":"nestjs-react-getting-started","region":"us-east-1","cloudProvider":"genezio-cloud","stage":"prod","stack":["nestjs"],"sourceRepository":"https://github.com/Genez-io/nestjs-react-getting-started","environmentVariables":[]}
✔ Changing spaces to tabs
09:09:15:498 DEBUG      deployCode.js:75        Response received {"status":"ok","projectId":"d4b3a8d9-b78c-42b9-8bef-f76a2ec59af8","projectEnvId":"237e134a-564f-492a-95d8-58b0617e2c41","createdAt":1740128164,"updatedAt":1740128955,"classes":[],"functions":[{"id":"2806dc59-b720-4bd0-8198-cf5b51bac16f","cloudUrl":"https://2806dc59-b720-4bd0-8198-cf5b51bac16f.dev-fkt.cloud.genez.io","name":"function-nest"}]}
09:09:15:501 DEBUG      utils.js:830    Creating archive of the project in /var/folders/gg/9f329j_x4m90bq72rxcyplv40000gn/T/genezio-42721/f0og3g
09:09:17:212 DEBUG      utils.js:74     No services found in the configuration file.
The app is being deployed at https://2806dc59-b720-4bd0-8198-cf5b51bac16f.dev-fkt.cloud.genez.io. It might take a few moments to be available worldwide.

App Dashboard URL: https://dev.app.genez.io/project/d4b3a8d9-b78c-42b9-8bef-f76a2ec59af8/237e134a-564f-492a-95d8-58b0617e2c41
Here you can monitor logs, set up a custom domain, and more.

09:09:17:214 DEBUG      command.js:22   Deploying classic genezio app
? We detected that SECRET are not set remotely. Do you want us to set them for you? (y/N)
```

in this case you detect because i have a frontend component

**Review by cristim67 — COMMENTED** [review]
i think you need to update de injection env logic.

tree:

```
.venv(base) cristimiloiu@MacBook-Pro-Cristi nestjs-react-getting-started % tree  -L 2 -a
.
├── backend
│   ├── ...
├── frontend
│   ├── ...
├── .env
└── genezio.yaml

17 directories, 26 files
```

logs:

```
✔ We detected that SECRET are not set remotely. Do you want us to set them for you? Yes
09:12:31:336 DEBUG      utils.js:643    Uploading environment variables [{"name":"SECRET","value":"test"}] from /Users/cristimiloiu/Desktop/workspace/work/nestjs-react-getting-started/.env.
09:12:31:986 DEBUG      utils.js:37     No services found in the configuration file.
⠸ Running frontend 1 deploy script09:12:32:195 DEBUG    genezio.js:148  Environment variables injected for frontend.scripts.deploy: {"VITE_API_URL_NESTJS":"https://2806dc59-b720-4bd0-8198-cf5b51bac16f.dev-fkt.cloud.genez.io"}
```

yaml:

```
frontend:
  # The folder where the frontend scripts will run.
  path: frontend
  # Specifies information about the SDK generation. If not specified, the SDK will not be generated for this frontend.
  # The directory that will be published to the CDN. It is relative to the `path` directory.
  publish: dist
  # Environment variables that will be injected into the frontend.
  environment:
    VITE_API_URL_NESTJS: ${{ nestjs.url }}
    VITE_API_ENV_VALUE: ${{env.SECRET}}
```

in this case you detect de env, but you don't inject the env.

**Review by cristim67 — COMMENTED** [review]
should handle the local flow too.

logs:

```

.venv(base) cristimiloiu@MacBook-Pro-Cristi nestjs-react-getting-started % genezio local
18:11:33:192 DEBUG      index.js:24     Sentry not initialized Cannot find module '/Users/cristimiloiu/Desktop/workspace/work/genezio/node_modules/@sentry/profiling-node/lib/sentry_cpu_profiler-darwin-arm64-131.node'
Require stack:
- /Users/cristimiloiu/Desktop/workspace/work/genezio/node_modules/@sentry/profiling-node/lib/index.js
Version: 3.2.2
18:11:34:413 DEBUG      local.js:1302   Starting SSR framework: NestJS
18:11:34:414 DEBUG      local.js:1303   SSR path: backend

 Error  Environment variable file was not provided or is not correct to be able to set ${{ env.SECRET }}. Please provide the correct path with `genezio deploy --env <envFile>`.

```

yaml:

```
nestjs:
  path: backend
  packageManager: npm
  environment:
    BACKEND_ENV: ${{ env.SECRET }}
```

**Review by cristim67 — COMMENTED** [review]
question: does this pull request also involve changes in the documentation?

**Review by cristim67 — COMMENTED** [review]
question: what is the workaround for a repo that is imported with github flow?

i think at second deploy you will raise an error because he don't have a '.env' file

File: src/commands/deploy/utils.ts [hunk 1]
```diff
@@ -932,95 +987,118 @@ export async function actionDetectedEnvFile(
...
 async function evaluateEnvironmentVariablesFromConfiguration(
     environment: Record<string, string>,
     configuration: YamlProjectConfiguration,
     stage: string,
-    envFile: string,
+    allowedResourceTypes: (
+        | "remoteResourceReference"
+        | "environmentFileReference"
+        | "literalValue"
+    )[],
```
Comment (cristim67) [inline src/commands/deploy/utils.ts:1092]: nit: create an enum or a type:

```
enum EnvironmentResourceType {
    RemoteResourceReference = "remoteResourceReference",
    EnvironmentFileReference = "environmentFileReference",
    LiteralValue = "literalValue"
}

allowedResourceTypes: EnvironmentResourceType[]
```

**Comment by andreia-oca** [comment]
> you should also handle the local flow.

Fixed.

> I think at deploy time, you will raise an error, because he don't have a .env file.

Afaik, on the Genezio CI/CD flow we are dynamically creating the `.env` file, so it will be okay. But great edge case, I will test it to see if it holds true.

> question: does this pull request also involve changes in the documentation?

I will silent the ${{ env.<KEY }} feature from the documentation until we define a good UX around it. They will remain in the codebase and we/our users can use them. For improvements/and an update, I create the following PR:
https://github.com/Genez-io/genezio-documentation/pull/316

**Review by cristim67 — APPROVED** [review]
